src/arch/riscv/isa.cc

```diff
@@ -331,6 +334,61 @@ ISA::setMiscReg(int misc_reg, RegVal val)
         warn("Ignoring write to %s.\n", CSRData.at(misc_reg).name);
     } else {
         switch (misc_reg) {
+
+          // From section 3.7.1 of RISCV priv. specs
+          // V1.12, the odd-numbered configuration
+          // registers are illegal for RV64 and
+          // each 64 bit CFG register hold configurations
+          // for 8 PMP entries.
+
+          case MISCREG_PMPCFG0:
+          case MISCREG_PMPCFG2:
+            {
+                // PMP registers should only be modified in M mode
+                assert(readMiscRegNoEffect(MISCREG_PRV) == PRV_M);
+
+                // Specs do not seem to mention what should be
+                // configured first, cfg or address regs!
+                // qemu seems to update the tables when
+                // pmp addr regs are written (with the assumption
+                // that cfg regs are already written)
+
+                // temp. printing for debugging
+                std::cout << "PMP cfg update " << std::endl;
+
+                int i;
+                for (i=0; i < sizeof(val); i++) {
+
+                    uint8_t cfgVal = (val >> 8 * i) & 0xff;
+                    auto mmu = dynamic_cast<RiscvISA::MMU *>
+                                (tc->getMMUPtr());
+                    mmu->getPMP()->pmpUpdateCfg
+                                (i+(4*(misc_reg-MISCREG_PMPCFG0)),cfgVal);
+                }
+
+                setMiscRegNoEffect(misc_reg, val);
+            }
+            break;
+          case MISCREG_PMPADDR00 ... MISCREG_PMPADDR15:
+            {
+                // temp. printing for debugging
+                std::cout << "PMP addr update " << std::endl;
+
+                // In qemu, the rule is updated whenever
+                // pmpaddr is written (in other words,
+                // pmpcfg/addr are converted) into a rule
+
+                // PMP registers should only be modified in M mode
+                assert(readMiscRegNoEffect(MISCREG_PRV) == PRV_M);
```

Comment:
An assert makes sense. Should we also (instead) raise an exception? Or, would an permission exception be raised somewhere else?

Reply:
Yeah, the assert was supposed to be there temporarily. It was not obvious to me from the specs that if/what exception should be raised. I will look again into it.